practice.rb

```diff
@@ -17,23 +43,56 @@ def pal(s)
 
 # Palindrome Tests
 raise "pal broke - pal('racecar')" unless pal("racecar") == true
+raise "pal broke - pal('deleveled')" unless pal("deleveled") == true
 raise "pal broke - pal('helloworld')" unless pal("helloworld") == false
 raise "pal broke - pal('')" unless pal("") == true
 puts "passes all palindrome tests"
 
+#Power Tests
+raise "power broke - power(2,3)" unless power(2,3) == 8
+raise "power broke - power(2,4)" unless power(2,4) == 16
+raise "power broke - power(9000,0)" unless power(9000,0) == 1
+raise "power broke - power(243,12)" unless power(243,3) == 14348907
+raise "power broke - power(2,4)" unless power(3,15) == 14348907
+puts "passes all power tests"
 
 
 
-# def binary(n)
-# end
-#
+def binary(n)
+	return "0 1" if n==1
+	previous_binary=binary(n-1)
+
+	#insert 0 into first position of each two copies of the previous binary string
+	#ex for binary(2), previous is "0 1", so  "0 1".dup.insert(0,"0") => "00 1"
+	#then substitute a 0 and a space for every space "00 1".gsub(/\s/, "0") => "00 01"
+	#now do this for 1s
+	# "0 1".dup.insert(0,"1") => "10 1"
+	# "10 1".gsub(/\s/, "1") => "10 11"
+
+	#Note:
+	#need .dup (duplicate) so that original string won't be modified by insert
+	#there is no non-destructive version of insert
+	first_half=previous_binary.dup.insert(0,"0").gsub(/\s/, " 0")
+	second_half=previous_binary.dup.insert(0,"1").gsub(/\s/, " 1")
+	return first_half+" "+ second_half
+end
```

Comment:
This one was tricky and weird since I didn't cover helper methods. Here's my solution:

```
def binary(n)
  binary_helper(n,"","")
end

def binary_helper(digits,binary,result)
  if digits == 0
    result += binary + " "
  else
    result = binary_helper(digits-1,binary+"0",result)
    result = binary_helper(digits-1,binary+"1",result)
  end
  return result
end
```